File: 000001-000200/000122.买卖股票的最佳时机II/122.js
/**
 * @param {number[]} prices
 * @return {number}
 */

 // 方法一：用时 68ms
 // 执行用时超过了 88%
 // 消耗内存超过了 18%
 // 难度: 中等
var maxProfit = function(prices) {
  let dp0 = new Array(prices.length).fill(0)
  let dp1 = new Array(prices.length).fill(0)
  dp0[0] = 0
  dp1[0] = -prices[0]

  for (let i = 1; i < prices.length; i++) {
    dp0[i] = Math.max(dp0[i - 1], dp1[i - 1] + prices[i])
    dp1[i] = Math.max(dp1[i - 1], dp0[i - 1] - prices[i])
  }
  // console.log('dp0:', JSON.stringify(dp0))
  // console.log('dp1:', JSON.stringify(dp1))
  return dp0[prices.length - 1]
};

console.log(maxProfit([7, 1, 5, 3, 6, 4]))
console.log(maxProfit([1, 2, 3, 4, 5]))

File: 000001-000200/000122.买卖股票的最佳时机II/122.go
package main

import (
	"fmt"
)

// 方法一：用时 4ms
// 执行用时超过了 87%
// 消耗内存超过了 20%
// 难度: 中等
func maxProfit(prices []int) int {
	dp0 := make([]int, len(prices))
	dp1 := make([]int, len(prices))
	dp1[0] = -prices[0]
	for i := 1; i < len(prices); i++ {
		dp0[i] = max(dp0[i - 1], dp1[i - 1] + prices[i])
		//fmt.Printf("i=%d dp0i=%d prices=%d dp0=%d\n", i, dp0[i - 1], prices[i], dp0[i])
		dp1[i] = max(dp1[i - 1], dp0[i - 1] - prices[i])
		//fmt.Printf("i=%d dp1i=%d prices=%d dp1=%d\n", i, dp1[i - 1], prices[i], dp1[i])
	}
	//fmt.Println("dp0:", dp0)
	//fmt.Println("dp1:", dp1)
	return dp0[len(prices) - 1]
}

func max (m int, n int) int {
	if m > n {
		return m
	}
	return n
}

func main ()  {
	fmt.Println(maxProfit([]int{7, 1, 5, 3, 6, 4}))
	fmt.Println(maxProfit([]int{1, 2, 3, 4, 5}))
}
